seal.go

```diff
@@ -0,0 +1,347 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package secboot
+
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"fmt"
+	"os"
+
+	"github.com/chrisccoulson/go-tpm2"
+
+	"golang.org/x/xerrors"
+)
+
+func computeSealedKeyDynamicAuthPolicy(tpm *tpm2.TPMContext, alg, signAlg tpm2.HashAlgorithmId, authKey *rsa.PrivateKey,
+	countIndexPub *tpm2.NVPublic, countIndexAuthPolicies tpm2.DigestList, pcrProfile *PCRProtectionProfile,
+	session tpm2.SessionContext) (*dynamicPolicyData, error) {
+	// Obtain the count for the new dynamic authorization policy
+	nextPolicyCount, err := readDynamicPolicyCounter(tpm, countIndexPub, countIndexAuthPolicies, session)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot read dynamic policy counter: %w", err)
+	}
+	nextPolicyCount += 1
+
+	countIndexName, _ := countIndexPub.Name()
+	if err != nil {
+		return nil, xerrors.Errorf("cannot compute name of dynamic policy counter: %w", err)
+	}
+
+	// Compute PCR digests
+	pcrValues, err := pcrProfile.computePCRValues(tpm, nil)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot compute PCR values from protection profile: %w", err)
+	}
+
+	// Use the PCR digests and NV index names to generate a single signed dynamic authorization policy digest
+	policyParams := dynamicPolicyComputeParams{
+		key:                  authKey,
+		signAlg:              signAlg,
+		pcrValues:            pcrValues,
+		policyCountIndexName: countIndexName,
+		policyCount:          nextPolicyCount}
+
+	policyData, err := computeDynamicPolicy(alg, &policyParams)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot compute dynamic authorization policy: %w", err)
+	}
+
+	return policyData, nil
+}
+
+// KeyCreationParams provides the arguments for SealKeyToTPM.
+type KeyCreationParams struct {
+	PinHandle tpm2.Handle // Handle at which to create a NV index for PIN support
```

Comment:
will this grow over time, it's a bit unclear why a struct?

should the pcrPolicy live in here?

Reply:
There's a good chance it will grow with additional features, such as additional 2nd factor mechanisms or if we add a network-bound element to unlocking. That said, moving pcrProfile here probably makes sense.

Reply:
Ok, I've added a PCRProfile field to KeyCreationParams.